tf_quant_finance/volatility/heston_approximation.py

```diff
@@ -0,0 +1,239 @@
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tensorflow as tf
+
+import math
+import tf_quant_finance as tff
+
+"""Method for semi-analytical Heston option price.
+
+Heston originally published in 1993 his eponymous model. He provided a semi-
+analytical formula for pricing European option via Fourrier transform under his
+model. However, as noted by Albrecher, the characteric function used in Heston 
+paper can suffer numerical issues because of the discontinuous nature of the 
+square root function in the complex plane, and a second version of the 
+characteric function which doesn't suffer this shortcoming should be used 
+instead. Attari further refined the numerical method by reducing the number of 
+numerical integrations (only one Fourrier transform instead of two) and with
+an integrand function with better convergence (converging in 1 / u **2 
+instead of 1 / u).  Attari's numerical method is implemented here.
+ 
+## References
+
+[1] Hansjorg Albrecher, The Little Heston Trap
+https://perswww.kuleuven.be/~u0009713/HestonTrap.pdf
+
+[2] Mukarram Attari, Option Pricing Using Fourier Transforms: A Numerically 
+Efficient Simplification
+https://papers.ssrn.com/sol3/papers.cfm?abstract_id=520042
+
+[3] Steven L. Heston, A Closed-Form Solution for Options with Stochastic 
+Volatility with Applications to Bond and Currency Options
+http://faculty.baruch.cuny.edu/lwu/890/Heston93.pdf
+"""
+
+def heston_price(*,
+                 strikes=None,
+                 expiries=None,
+                 is_call_options=None,
+                 kappas=None,
+                 thetas=None,
+                 sigmas=None,
+                 rhos=None,
+                 variances=None,
+                 forwards=None,
+                 discount_factors=None,
+                 dtype=None,
+                 name=None):
+  """Calculates European option prices under the Heston model.
+
+  This uses the numerical method proposed by Attari in 2004.
+  Heston model:
+    dF/F = sqrt(V) * dW_1
+    dV = kappa * (theta - V) * dt * sigma * sqrt(V) * dW_2
+    <dW_1,dW_2> = rho *dt
+  The variance V follows a square root process.
+
+  #### Example
+
+  ```python
+  prices = heston_price(
+    kappas =  np.asarray([[2.0]]),
+    thetas=np.asarray([[0.5]]),
+    sigmas=np.asarray([[0.15]]),
+    rhos=np.asarray([[0.3]]),
+    variances=np.asarray([[0.11]]),
+    forwards=np.asarray([[100.0]]),
+    expiries=np.asarray([[1.2]]),
+    is_call_options=np.asarray([True], dtype=np.bool),
+    strikes=np.asarray([[102.0]]),
+    discount_factors=np.asarray([[1.0]]),
+  )
+
+  # Expected print output of prices:
+  # [24.82219619]
```

Comment:
Why is the result not of the same shape as kappas?

Reply:
good point, the docstring was incorrect